src/java/org/apache/cassandra/index/sai/disk/hnsw/OnDiskOrdinalsMap.java

```diff
@@ -24,72 +24,84 @@
 
-import org.apache.cassandra.io.util.File;
-import org.apache.cassandra.io.util.RandomAccessReader;
+import org.apache.cassandra.io.util.FileHandle;
 
 public class OnDiskOrdinalsMap
 {
     private static final Logger logger = LoggerFactory.getLogger(OnDiskOrdinalsMap.class);
 
-    private final RandomAccessReader reader;
+    private final FileHandle fh;
     private final int size;
     private final long rowOrdinalOffset;
 
-    public OnDiskOrdinalsMap(File file) throws IOException
+    public OnDiskOrdinalsMap(FileHandle fh)
     {
-        this.reader = RandomAccessReader.open(file);
-        this.size = reader.readInt();
-        reader.seek(reader.length() - 8);
-        this.rowOrdinalOffset = reader.readLong();
+        this.fh = fh;
+        try (var reader = fh.createReader())
+        {
+            this.size = reader.readInt();
+            reader.seek(reader.length() - 8);
+            this.rowOrdinalOffset = reader.readLong();
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
     }
 
     public int[] getSegmentRowIdsMatching(int vectorOrdinal) throws IOException
     {
         Preconditions.checkArgument(vectorOrdinal < size, "vectorOrdinal %s is out of bounds %s", vectorOrdinal, size);
 
-        // read index entry
-        reader.seek(4L + vectorOrdinal * 8L);
-        var offset = reader.readLong();
-        // seek to and read ordinals
-        reader.seek(offset);
-        var postingsSize = reader.readInt();
-        var ordinals = new int[postingsSize];
-        for (var i = 0; i < ordinals.length; i++)
+        try (var reader = fh.createReader())
         {
-            ordinals[i] = reader.readInt();
+            // read index entry
+            reader.seek(4L + vectorOrdinal * 8L);
+            var offset = reader.readLong();
+            // seek to and read ordinals
+            reader.seek(offset);
+            var postingsSize = reader.readInt();
+            var ordinals = new int[postingsSize];
+            for (var i = 0; i < ordinals.length; i++)
+            {
+                ordinals[i] = reader.readInt();
+            }
+            return ordinals;
         }
-        return ordinals;
     }
 
     /**
      * @return order if given row id is found; otherwise return -1
      */
     public int getOrdinalForRowId(int rowId) throws IOException
     {
-        // Compute the offset of the start of the rowId to vectorOrdinal mapping
-        var high = (reader.length() - 8 - rowOrdinalOffset) / 8;
-        long index = DiskBinarySearch.searchInt(0, Math.toIntExact(high), rowId, i -> {
-            try
-            {
-                long offset = rowOrdinalOffset + i * 8;
-                reader.seek(offset);
-                return reader.readInt();
-            }
-            catch (IOException e)
-            {
-                throw new RuntimeException(e);
-            }
-        });
+        try (var reader = fh.createReader())
```

Comment:
I know I said it wasn't that expensive but isn't there a way we could create te `OnDiskOrdinalsMap` per query during searches? It probably doesn't matter immediately to test for correctness but it's definitely something we could ticket to look at later.